From 8ef32bbdbc476e3f6d31a457d03029b644e6145b Mon Sep 17 00:00:00 2001
From: Arulanbu Balusamy <quic_abalusam@quicinc.com>
Date: Tue, 3 Sep 2024 18:05:19 +0530
Subject: [PATCH] hostapd: Fix beacon rate configuration for non-legacy rates.

Prevent assigning NL80211_TXRATE_VHT while configuring the
beacon rates for 6Ghz. Because VHT rates are unsupported for
6GHz frequencies.

Duplicate beacons are only supported for non-HT rates. Therefore,
set the HE_6GHZ_OPER_INFO_CTRL_DUP_BEACON bit exclusively for legacy
rates.

Patch-dependency:  [None]
Patch-work:        [None]

Signed-off-by: Arulanbu Balusamy <quic_abalusam@quicinc.com>
---
 src/ap/ieee802_11_he.c       | 3 ++-
 src/drivers/driver_nl80211.c | 8 +++++---
 2 files changed, 7 insertions(+), 4 deletions(-)

--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -287,7 +287,8 @@ u8 * hostapd_eid_he_operation(struct hos
 
 		if (center_idx_to_bw_6ghz(seg0) &&
 		    !is_6ghz_psc_frequency(ieee80211_chan_to_freq(NULL,
-				    hapd->iconf->op_class, hapd->iconf->channel)))
+				    hapd->iconf->op_class, hapd->iconf->channel)) &&
+				    hapd->iconf->rate_type == BEACON_RATE_LEGACY)
 			control |= HE_6GHZ_OPER_INFO_CTRL_DUP_BEACON;
 
 		*pos++ = control;
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -4650,8 +4650,9 @@ static int nl80211_put_beacon_rate(struc
 		he_rate.mcs[0] = BIT(params->beacon_rate);
 		if (nla_put(msg, NL80211_TXRATE_LEGACY, 0, NULL) ||
 		    nla_put(msg, NL80211_TXRATE_HT, 0, NULL) ||
+		    (params->freq->vht_enabled &&
 		    nla_put(msg, NL80211_TXRATE_VHT, sizeof(vht_rate),
-			    &vht_rate) ||
+			    &vht_rate)) ||
 		    nla_put(msg, NL80211_TXRATE_HE, sizeof(he_rate), &he_rate))
 			return -1;
 		wpa_printf(MSG_DEBUG, " * beacon_rate = HE-MCS %u",
@@ -4667,8 +4668,9 @@ static int nl80211_put_beacon_rate(struc
 		eht_rate.mcs[0] = BIT(params->beacon_rate);
 		if (nla_put(msg, NL80211_TXRATE_LEGACY, 0, NULL) ||
 		    nla_put(msg, NL80211_TXRATE_HT, 0, NULL) ||
-		    nla_put(msg, NL80211_TXRATE_VHT, sizeof(vht_rate),
-			    &vht_rate) ||
+		    (params->freq->vht_enabled &&
+		     nla_put(msg, NL80211_TXRATE_VHT, sizeof(vht_rate),
+			     &vht_rate)) ||
 		    nla_put(msg, NL80211_TXRATE_HE, sizeof(he_rate),
 			    &he_rate) ||
 		    nla_put(msg, NL80211_TXRATE_EHT, sizeof(eht_rate),
